network-mux/src/Network/Mux/Trace.hs

```diff
@@ -0,0 +1,157 @@
+{-# LANGUAGE ExistentialQuantification #-}
+{-# LANGUAGE NamedFieldPuns            #-}
+{-# LANGUAGE DeriveFunctor             #-}
+
+module Network.Mux.Trace (
+      MuxError(..)
+    , MuxErrorType(..)
+    , MuxTrace(..)
+    , MuxBearerState(..)
+    , WithMuxBearer(..)
+    , TraceLabelPeer(..)
+    ) where
+
+import           Prelude hiding (read)
+
+import qualified Data.ByteString.Lazy as BL
+import           Text.Printf
+
+import           Control.Exception
+import           GHC.Stack
+
+import           Network.Mux.Types
+
+
+--
+-- Errors
+--
+
+-- | Error type used in accross the mux layer.
+--
+data MuxError = MuxError {
+      errorType  :: !MuxErrorType
+    , errorMsg   :: !String
+    , errorStack :: !CallStack
+    } deriving Show
+
+
+-- | Enumeration of error conditions.
+--
+data MuxErrorType = MuxUnknownMiniProtocol
+                  -- ^ returned by 'decodeMuxSDUHeader', thrown by 'MuxBearer'.
+                  | MuxDecodeError
+                  -- ^ return by 'decodeMuxSDUHeader', thrown by 'MuxBearer'.
+                  | MuxBearerClosed
+                  -- ^ thrown by 'MuxBearer' when received a null byte.
+                  | MuxIngressQueueOverRun
+                  -- ^ thrown by 'demux' when violating 'maximumIngressQueue'
+                  -- byte limit.
+                  | MuxControlProtocolError
+                  -- ^ thrown by 'muxControl' (mux control thread), when
+                  -- received a 'Muxcontrol' message on a mature 'MuxBearer'.
+                  |  MuxTooLargeMessage
+                  -- ^ thrown by 'muxChannel' when violationg
+                  -- 'maximumMessageSize' byte limit.
+                  deriving (Show, Eq)
+
+instance Exception MuxError where
+    displayException MuxError{errorType, errorMsg, errorStack}
+      = printf "%s %s at %s"
+          (show errorType)
+          (show errorMsg)
+          (prettyCallStack errorStack)
+
+
+--
+-- Tracing
+--
+
+-- | A peer label for use in 'Tracer's. This annotates tracer output as being
+-- associated with a given peer identifier.
+--
+data TraceLabelPeer peerid a = TraceLabelPeer peerid a
+  deriving (Eq, Functor, Show)
+
+-- | Type used for tracing mux events.
+--
+data WithMuxBearer peerid a = WithMuxBearer {
+      wmbPeerId :: !peerid
+      -- ^ A tag that should identify a specific mux bearer.
+    , wmbEvent  :: !a
+}
+--TODO: probably remove this type
```

Comment:
Why?

Reply:
Oh, we have an equivalent one in `ouroboros-network` and I think that's probably the better place to have it. At the level of mux we only have a single peer anyway, so it's unnecessary to have the concept of a peer id. It's only when we have multiple that we need to tag them, and that's done in `ouroboros-network`.